icefoganalytics/sfa-client
# Messaging Module: Student Training Allowance Letters

Fixes https://github.com/icefoganalytics/sfa-client/issues/57

# Context

Add new letter templates.
- Yukon University - student training allowance approval
- non-Yukon University student training allowance approval
- student training allowance rejection

Initial work began in https://github.com/icefoganalytics/sfa-client/commit/bffcdb7d4f8f57867799af9edc6124480b6019f8

After attempting to find valid data for quite some time, and failing. I want to push up the priority of https://github.com/icefoganalytics/sfa-client/issues/52. We might also want to build out a Mailer Preview feature. i.e. https://guides.rubyonrails.org/action_mailer_basics.html#previewing-emails. With some data factories its pretty easy to do.

# Implementation

I opted to keep the letters specific and divergent despite the high degree of overlap. I think this is the best choice for now as it allows the letters to change independently. 

If we seem to be using a lot of shared template logic we should use some template partials. 

If we use a lot of shared serialization, we should move some of these methods to a "base serializer" class.

## TODO

- [ ] QA Student Training Allowance - Yukon University letter
- [ ] QA Student Training Allowance - Alkan Air letter

## Questions

- [ ] should I limit the letters returned by /api/v2/admin/funding-requests/:fundingRequestId/letters by request status and (optionally) institution name? The front-end should be able to provide the appropriate slug, but it might be expedient to return only the valid slugs for a given funding request? This would require that the Student Training Allowance letters would also need to discriminate on the institution name. It would be easy to do in the FundingRequestsLettersController#listLetters code, or maybe with an extra service.

# Screenshots

Yukon Training Allowance letter metadata
![image](https://github.com/icefoganalytics/sfa-client/assets/23045206/8c490093-5de5-4386-a75f-d3a033fced53)

Yukon Training Allowance Rejection
![image](https://github.com/icefoganalytics/sfa-client/assets/23045206/6989f701-2028-40c0-ba2b-b06a8c84c96a)


# Testing Instructions

1. Boot the api and db services via `dev up`.
2. Boot the web service via `cd src/web && npm run start`
4. Log in to the web UI at http://localhost:8080.
5. Check that you can see all the current letter options by going to http://localhost:3000/api/v2/admin/funding-request-letters. You will see three new letter options: 
    i. Student Training Allowance - Yukon University
    ii. Student Training Allowance - Alkan Air
    iii. Student Training Allowance Rejection
6. Check that you can see the specific letters that match the Yukon Training Allowance request type at http://localhost:3000/api/v2/admin/funding-requests/250/letters.
7. Check that the new Student Training Allowance Rejection renders correctly at http://localhost:3000/api/v2/admin/funding-requests/1122/letters/student-training-allowance-rejection.pdf
8. Check that the Student Training Allowance - Yukon University letter renders correctly at http://localhost:3000/api/v2/admin/funding-requests/:fundingRequestId/letters/student-training-allowance-yukon-university-approval.pdf
> TODO: find a valid funding request id, most are missing address information
9. Check that the Student Training Allowance - Yukon University letter renders correctly at http://localhost:3000/api/v2/admin/funding-requests/:fundingRequestId/letters/student-training-allowance-alkan-air-approval.pdf
> TODO: find a valid funding request id, most are missing address information
